chore: Refactor `rover new` into `rover template` (#1299)
apollographql/apollo-cli: docs/source/commands/template.mdx

---
title: Rover template commands
description: For use with Apollo Federation
---

A **supergraph** (also called a federated graph) is a graph composed of multiple **subgraphs**:

```mermaid
graph BT;
  gateway(["Supergraph (A + B + C)"]);
  serviceA[Subgraph A];
  serviceB[Subgraph B];
  serviceC[Subgraph C];
  gateway --- serviceA & serviceB & serviceC;
```

Rover template commands help you create new subgraphs from templated projects in various [federation-compatible implementations](/federation/supported-subgraphs).

This feature helps you quickly start up a new GraphQL project or explore various subgraph implementations. If you want to contribute a new template, [open an issue here](https://github.com/apollographql/rover/issues/new/choose)!

## Create a new subgraph

Run `rover template use <PATH>` to generate the boilerplate for a new subgraph. You can either use the guided wizard experience to select a template, or specify a template with the `--template` flag.

> ### Using the Wizard

Reviewer: ```suggestion
### Using the wizard
```

---
title: Rover template commands
description: For use with Apollo Federation
---

A **supergraph** (also called a federated graph) is a graph composed of multiple **subgraphs**:

```mermaid
graph BT;
  gateway(["Supergraph (A + B + C)"]);
  serviceA[Subgraph A];
  serviceB[Subgraph B];
  serviceC[Subgraph C];
  gateway --- serviceA & serviceB & serviceC;
```

> Rover template commands help you create new subgraphs from templated projects in various [Federation-compatible implementations](/federation/supported-subgraphs).

Reviewer: ```suggestion
Rover template commands help you create new subgraphs from templated projects in various [federation-compatible implementations](/federation/supported-subgraphs).
```

> This feature is intended to help quickly start up a new GraphQL project and enable new developers to explore various implementations. If you have an idea or want to contribute a specific template, open an issue here!

Reviewer: ```suggestion
This feature helps you quickly start up a new GraphQL project or explore various subgraph implementations. If you want to contribute a new template, open an issue here!
```

Reviewer: Also presumably need a link for the "open an issue here" part

## Using the wizard to create a new subgraph project

> You can use `rover template use` to start a wizard that will help you create a new subgraph project.

Reviewer: ```suggestion
Run `rover template use` to start a wizard that helps you create a new subgraph project. The wizard walks you through a series of questions to help you choose a template.
```

> 
> > You will be prompted through a series of questions to filter down to the exact template you want to use.

Reviewer: ```suggestion
```

> You can optionally provide a `--path` option to specify which relative path to create the project in. If you do not provide a path, the project will be created in a new directory named the ID of the template.

Reviewer: ```suggestion
You can optionally provide a `--path` option to specify which relative path to create your project in. If you don't provide a path, the project is created in a new directory with a name that matches the ID of the template.
```

```bash
rover template use --path <PATH>
```

### Project types

> You will be prompted to select which type of project you want to generate. You can optionally provide a `--project-type` option to skip this prompt.

Reviewer: ```suggestion
By default, the wizard prompts you to select which type of project you want to generate. You can optionally provide a `--project-type` option to skip this prompt.
```

> Run `rover template use --help` to see a list of available project types.

```shell
rover template use --path="my-new-subgraph" --project-type=subgraph
```

> ### Project Languages

Reviewer: ```suggestion
### Project languages
```

> You will be prompted to choose a programming language for your project. You can optionally provide a `--language` option to skip this prompt.

Reviewer: ```suggestion
By default, the wizard prompts you for which programming language to use for your project. You can optionally provide a `--language` option to skip this prompt.
```

> Run `rover template use --help` to see a list of available languages.

```shell
rover template use --path="my-new-python-subgraph" --language=python
```

> ## Skipping the wizard

Reviewer: ```suggestion
### Skipping the wizard
```

Author: This was intended to be at the same level as the previous "Using the wizard to create a new subgraph project" since this is a different path forward (presumably for users who are already familiar with the templates).

If you know exactly which template you want to use, you can specify the `--template` option to skip the wizard:

```shell
rover template use --path="my-new-apollo-server-subgraph" --template="subgraph-javascript-apollo-server"
```

> Use `rover template list` to see a list of available templates.

## Listing available templates

> `rover template list` will list all available templates with their ID (for use with `--template` argument to `rover template use`), name, project type, language, and Repo URL. To learn more about a specific template, go to the Repo URL and read the template's `README.md`.

Reviewer: ```suggestion
The `rover template list` command lists each available template with its corresponding ID (which you can provide to the `--template` option of `rover template use`). It also displays each template's name, project type, language, and Repo URL. To learn more about a specific template, go to its repo URL and see its `README.md` file.
```
